### PR TITLE
Respect child specified size in CKStackLayoutComponent

### PR DESCRIPTION
**Pull request by ocrickard**
I need to finish up recording the snapshot tests (ugh), but this demonstrates an issue we have in stack layout.

### CODE REVIEW

**Comment by ocrickard** [comment]
Alright @eczarny mind giving this a quick review to see how you feel about the implementation/ask any clarifying questions we need here? I'm going to put up an internal version of this to run snapshot tests on. I'm pretty sure it's going to break *everything*.

By Monday I'll add more snapshot tests for all the different edge cases to this diff, and we can discuss a landing plan.

File: ComponentKit/LayoutComponents/CKStackUnpositionedLayout.mm [hunk 1]
```diff
@@ -15,10 +15,38 @@
...
+ 
+ */
+static CGFloat resolveCrossDimensionMaxForStretchChild(const CKStackLayoutComponentStyle &style,
+                                                       const CKStackLayoutComponentChild &child,
+                                                       const CGFloat stackMax,
+                                                       const CGFloat crossMax)
+{
+  // stretched children may have a cross direction max that is smaller than the minimum size constraint of the parent.
+  const CGFloat computedMax = (style.direction == CKStackLayoutDirectionVertical ?
+                               child.component.size.resolve(kCKComponentParentSizeUndefined).max.width :
```
Comment (eczarny) [inline ComponentKit/LayoutComponents/CKStackUnpositionedLayout.mm]: What are the implications of resolving with an undefined parent size?

File: ComponentKit/LayoutComponents/CKStackUnpositionedLayout.mm [hunk 2]
```diff
@@ -15,10 +15,38 @@
...
+                               child.component.size.resolve(kCKComponentParentSizeUndefined).max.height);
+  return computedMax == INFINITY ? crossMax : computedMax;
```
Comment (eczarny) [inline ComponentKit/LayoutComponents/CKStackUnpositionedLayout.mm]: Interesting. So if the computed maximum is infinity then we should stretch child? Otherwise use the computed maximum?

**Comment by eczarny** [comment]
Woah. GitHub can do proper review now?

File: ComponentKit/LayoutComponents/CKStackUnpositionedLayout.mm (hunk 1 above)
Comment (ocrickard) [inline ComponentKit/LayoutComponents/CKStackUnpositionedLayout.mm]: It allows us to avoid trying to resolve things like a relative percentage, since that should explicitly not be defined here.

File: ComponentKit/LayoutComponents/CKStackUnpositionedLayout.mm (hunk 2 above)
Comment (ocrickard) [inline ComponentKit/LayoutComponents/CKStackUnpositionedLayout.mm]: If it's infinity, then it is unconstrained, in which case we can properly stretch the item, and directly use the crossMax size, yes.

**Comment by ocrickard** [comment]
@facebook-github-bot import

**Comment by facebook-github-bot-2** [comment]
Thanks for importing. If you are an FB employee go to [Phabricator](https://phabricator.intern.facebook.com/D3886493) to review internal test results.

**Comment by facebook-github-bot-9** [comment]
@ocrickard updated the pull request - [view changes](https://github.com/facebook/componentkit/pull/620/files/514f22de0037e892ac7c191f0d23171fc2b44bbf..d41a68124e17e786caa7e002e584527961832f8b) - [changes since last import](https://github.com/facebook/componentkit/pull/620/files/514f22de0037e892ac7c191f0d23171fc2b44bbf..d41a68124e17e786caa7e002e584527961832f8b)

**Comment by facebook-github-bot-6** [comment]
@ocrickard updated the pull request - [view changes](https://github.com/facebook/componentkit/pull/620/files/d41a68124e17e786caa7e002e584527961832f8b..ab51f167607709b7e56fc0e96176a4277898b347) - [changes since last import](https://github.com/facebook/componentkit/pull/620/files/514f22de0037e892ac7c191f0d23171fc2b44bbf..ab51f167607709b7e56fc0e96176a4277898b347)

**Comment by eczarny** [comment]
@ocrickard is this ready for review?

**Comment by ocrickard** [comment]
@eczarny yep! I was thinking about addding more tests, but don't want to go overboard here. Open to suggestions on what else we should add.

File: ComponentKit/LayoutComponents/CKStackUnpositionedLayout.mm [hunk 3]
```diff
@@ -15,10 +15,38 @@
 #import <ComponentKit/CKAssert.h>
 
 #import "ComponentUtilities.h"
+#import "CKComponentInternal.h"
 #import "CKComponentSubclass.h"
 #import "CKStackLayoutComponentUtilities.h"
 
 /**
+ 
+ */
```
Comment (eczarny) [inline ComponentKit/LayoutComponents/CKStackUnpositionedLayout.mm]: Plan on adding anything here?

File: ComponentKit/LayoutComponents/CKStackUnpositionedLayout.mm [hunk 4]
```diff
@@ -15,10 +15,38 @@
...
+static CGFloat resolveCrossDimensionMinForStretchChild(const CKStackLayoutComponentStyle &style,
+                                                    const CKStackLayoutComponentChild &child,
+                                                    const CGFloat stackMax,
+                                                    const CGFloat crossMin)
+{
+  // stretched children will have a cross dimension of at least crossMin, unless they explicitly define a child size
+  // that is smaller than the constraint of the parent.
+  return (style.direction == CKStackLayoutDirectionVertical ?
+             child.component.size.resolve(kCKComponentParentSizeUndefined).min.width :
+          child.component.size.resolve(kCKComponentParentSizeUndefined).min.height) ?: crossMin;
```
Comment (eczarny) [inline ComponentKit/LayoutComponents/CKStackUnpositionedLayout.mm]: Does the `?:` operator work? I thought it was specifically for `nil` values but perhaps it also works with any zero-like value?

Bit of a nit: the indentation feels a bit off here. Should line #45 be at the same indentation level as this line?

File: ComponentKit/LayoutComponents/CKStackUnpositionedLayout.mm (hunk 3 above)
Comment (ocrickard) [inline ComponentKit/LayoutComponents/CKStackUnpositionedLayout.mm]: lol, oops. Yeah.

File: ComponentKit/LayoutComponents/CKStackUnpositionedLayout.mm (hunk 4 above)
Comment (ocrickard) [inline ComponentKit/LayoutComponents/CKStackUnpositionedLayout.mm]: It does work, and yeah agreed on the indentation. Not sure what happened there.

**Comment by ocrickard** [comment]
I'll merge once CI finishes, then push the 0.16 tag tomorrow morning.

**Comment by ocrickard** [comment]
Alright, CI finished, I'm merging. Here's to hoping things don't dramatically break... we'll revert if so.
